src/NHibernate/Multi/QueryBatchItemBase.cs

```diff
@@ -171,80 +177,96 @@ public int ProcessResultsSet(DbDataReader reader)
 			var dialect = Session.Factory.Dialect;
 			var hydratedObjects = new List<object>[_queryInfos.Count];
 
-			var rowCount = 0;
-			for (var i = 0; i < _queryInfos.Count; i++)
+			var cacheModeOrig = Session.CacheMode;
+			if (_cacheMode.HasValue &&
+				// Avoid setting the cache mode with the same value as the setter is not just affecting
+				// the backing field
+				_cacheMode != cacheModeOrig)
 			{
-				var queryInfo = _queryInfos[i];
-				var loader = queryInfo.Loader;
-				var queryParameters = queryInfo.Parameters;
-
-				//Skip processing for items already loaded from cache
-				if (queryInfo.IsResultFromCache)
+				Session.CacheMode = _cacheMode.Value;
```

Comment:
As this code is used more than once and it's not really one-liner I would wrap it in `IDisposable` class. Something like:
```C#
class SessionContext : IDisposable
{
	private readonly ISession _session;
	private readonly CacheMode? _cacheModeToReturn;

	public SessionContext(ISession session, CacheMode? cacheMode)
	{
		if (cacheMode.HasValue && cacheMode != session.CacheMode)
		{
			_cacheModeToReturn = session.CacheMode;
			_session.CacheMode = cacheMode.Value;
		}
	}

	public void Dispose()
	{
		if(_cacheModeToReturn.HasValue)
			_session.CacheMode = _cacheModeToReturn.Value;
	}
}
```
So usage in code would be more concise:
```C#
using(new SessionContext(Session, _cacheMode)
...
```

I think  this class can also be made private in `QueryParameters` and exposed only via method. Something like: 
```C#
using(queryParams.FillSessionContext(session))
```
So it can be extended easily with filling some other context fields if added. But not sure if it's really usefull.

Reply:
Agree.